matrix/matrix.py

```diff
@@ -0,0 +1,184 @@
+"""Matrix data type class."""
+
+import operator
+from collections.abc import Iterable
+
+
+class Matrix:
+    """An immutable object representing a matrix data-type.
+
+    Attributes:
+        _matrix (tuple): contains the data about the matrix. Each line of
+            the matrix is represented by a tuple.
+    Note:
+        This implementation only supports squared matrices (n * n).
+    """
+
+    def __init__(self, *lines):
+        self._validate(*lines)
+        self._matrix = tuple(lines)
+
+    def _validate(self, *data: tuple):
+        """Check if the data given is valid.
+
+        Note:
+            Valid data is defined if it contains tuples of the same length
+            (Matrix only supports n * n), items can't be iterable.
+        """
+        if not all(isinstance(line_data, tuple) for line_data in data):
+            raise ValueError("Matrix lines must be tuple")
+
+        if any(isinstance(item, Iterable) for line_data in data for item in
```

Comment:
Remove all isinstance usage, the Pythonic way is to duck-type everything

The only place you must use isinstance is the multiplication